File: api/src/test/java/ru/endlesscode/mimic/api/system/LevelSystemTest.java
/*
 * This file is part of MimicAPI.
 * Copyright (C) 2018 Osip Fatkullin
 * Copyright (C) 2018 EndlessCode Group and contributors
 *
 * MimicAPI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * MimicAPI is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with MimicAPI.  If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.mimic.api.system;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.*;

public class LevelSystemTest {
    private LevelSystem ls;

    @Before
    public void setUp() {
        this.ls = new BasicLevelSystemImpl();
    }

    @Test
    public void testGettingConverter() {
        ExpLevelConverter converter = ls.getConverter();

        assertNotNull(converter);
        assertTrue(converter instanceof BasicConverterImp);
    }

    @Test
    public void testLevelGetSet() {
        this.ls.setLevel(10);
        assertEquals(10, ls.getLevel());

        this.ls.setLevel(-10);
        assertEquals(0, ls.getLevel());
    }

    @Test
    public void testExpGetSet() {
        this.ls.setExp(9);
        assertEquals(9, ls.getExp());

        this.ls.setExp(10);
        assertEquals(9, ls.getExp());

        this.ls.setExp(0);
        assertEquals(0, ls.getExp());
    }

    @Test
    public void testIncreaseLevel() {
        this.ls.setLevel(1);
        this.ls.increaseLevel(4);

        assertEquals(5, this.ls.getLevel());
    }

    @Test
    public void testDecreaseLevel() {
        this.ls.setLevel(10);

        this.ls.decreaseLevel(6);
        assertEquals(4, this.ls.getLevel());

        this.ls.decreaseLevel(6);
        assertEquals(0, this.ls.getLevel());
    }

    @Test
    public void testDidReachLevel() {
        this.ls.setLevel(10);

        assertTrue(this.ls.didReachLevel(0));
        assertTrue(this.ls.didReachLevel(10));
        assertFalse(this.ls.didReachLevel(11));
    }

    @Test
    public void testGetFractionalExp() {
        this.ls.setExp(-5);
        assertEquals(0., this.ls.getFractionalExp(), 0.001);

        this.ls.setExp(0);
        assertEquals(0., this.ls.getFractionalExp(), 0.001);

        this.ls.setExp(5);
        assertEquals(.5, this.ls.getFractionalExp(), 0.001);

        this.ls.setExp(9);
        assertEquals(.9, this.ls.getFractionalExp(), 0.001);

        this.ls.setExp(15);
        assertEquals(.9, this.ls.getFractionalExp(), 0.001);
    }

    @Test
    public void testSetFractionalExp() {
        this.ls.setFractionalExp(-1);
        assertEquals(0, this.ls.getExp());

        this.ls.setFractionalExp(0);
        assertEquals(0, this.ls.getExp());

        this.ls.setFractionalExp(0.5123);
        assertEquals(5, this.ls.getExp());

        this.ls.setFractionalExp(1);
        assertEquals(9, this.ls.getExp());

        this.ls.setFractionalExp(2);
        assertEquals(9, this.ls.getExp());
    }

    @Test
    public void testGetTotalExp() {
        this.ls.setLevel(0);
        this.ls.setExp(0);
        assertEquals(0, this.ls.getTotalExp());

        this.ls.setLevel(5);
        this.ls.setExp(0);
        assertEquals(50, this.ls.getTotalExp());

        this.ls.setLevel(0);
        this.ls.setExp(5);
        assertEquals(5, this.ls.getTotalExp());

        this.ls.setLevel(4);
        this.ls.setExp(9);
        assertEquals(49, this.ls.getTotalExp());
    }

    @Test
    public void testSetTotalExp() {
        this.ls.setTotalExp(-10);
        assertEquals(0, this.ls.getLevel());
        assertEquals(0, this.ls.getExp());

        this.ls.setTotalExp(0);
        assertEquals(0, this.ls.getLevel());
        assertEquals(0, this.ls.getExp());

        this.ls.setTotalExp(50);
        assertEquals(5, this.ls.getLevel());
        assertEquals(0, this.ls.getExp());

        this.ls.setTotalExp(5);
        assertEquals(0, this.ls.getLevel());
        assertEquals(5, this.ls.getExp());

        this.ls.setTotalExp(49);
        assertEquals(4, this.ls.getLevel());
        assertEquals(9, this.ls.getExp());
    }

    @Test
    public void testHasExp() {
        this.ls.setLevel(1);
        this.ls.setExp(5);

        assertTrue(this.ls.hasExp(15));
        assertTrue(this.ls.hasExp(0));
        assertFalse(this.ls.hasExp(16));
    }

    @Test
    public void testGiveExp() {
        this.ls.setLevel(10);
        this.ls.setExp(6);

        this.ls.giveExp(3);
        assertEquals(10, this.ls.getLevel());
        assertEquals(9, this.ls.getExp());

        this.ls.giveExp(1);
        assertEquals(11, this.ls.getLevel());
        assertEquals(0, this.ls.getExp());

        this.ls.giveExp(15);
        assertEquals(12, this.ls.getLevel());
        assertEquals(5, this.ls.getExp());
    }

    @Test
    public void testTakeExp() {
        this.ls.setLevel(10);
        this.ls.setExp(6);

        this.ls.takeExp(6);
        assertEquals(10, this.ls.getLevel());
        assertEquals(0, this.ls.getExp());

        this.ls.takeExp(1);
        assertEquals(9, this.ls.getLevel());
        assertEquals(9, this.ls.getExp());

        this.ls.takeExp(16);
        assertEquals(8, this.ls.getLevel());
        assertEquals(3, this.ls.getExp());

        this.ls.takeExp(100);
        assertEquals(0, this.ls.getLevel());
        assertEquals(0, this.ls.getExp());
    }
}

File: mimic-api/src/test/kotlin/level/LevelSystemTest.kt
/*
 * This file is part of Mimic.
 * Copyright (C) 2020 Osip Fatkullin
 * Copyright (C) 2020 EndlessCode Group and contributors
 *
 * Mimic is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Mimic is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Mimic.  If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.mimic.level

import io.kotest.matchers.shouldBe
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import kotlin.test.Test

@Suppress("DEPRECATION")
class LevelSystemTest {

    class DummyLevelSystem : LevelSystem {
        override val converter: ExpLevelConverter = DummyExpLevelConverter()
        override var level: Int = 0
        override var exp: Double = 0.0
    }

    // SUT
    private val levelSystem: LevelSystem = DummyLevelSystem()

    @Test
    fun `when increase level - should set right level`() {
        levelSystem.level = 1
        levelSystem.giveLevels(4)
        levelSystem.level shouldBe 5
    }

    @ParameterizedTest
    @CsvSource(
        "4, 2",
        "10, 0",
    )
    fun `test takeLevels`(lvlAmount: Int, expectedLevel: Int) {
        levelSystem.level = 6
        levelSystem.takeLevels(lvlAmount)
        levelSystem.level shouldBe expectedLevel
    }

    @ParameterizedTest
    @CsvSource(
        "5,     true",
        "-1,    true",
        "10,    true",
        "11,    false",
    )
    fun `test didReachLevel when level is 10`(requiredLevel: Int, isReached: Boolean) {
        levelSystem.level = 10
        levelSystem.didReachLevel(requiredLevel) shouldBe isReached
    }

    @ParameterizedTest
    @CsvSource(
        "-1.0,  0.0",
        "0.0,   0.0",
        "5.12,  0.512",
        "11.0,  0.9"
    )
    fun `test exp to fractionalExp`(exp: Double, fractionalExp: Double) {
        levelSystem.exp = exp
        levelSystem.fractionalExp shouldBe fractionalExp
    }

    @ParameterizedTest
    @CsvSource(
        "-1.0,  0.0",
        "0.0,   0.0",
        "0.512, 5.12",
    )
    fun `test fractionalExp to exp`(fractionalExp: Double, exp: Double) {
        levelSystem.fractionalExp = fractionalExp
        levelSystem.exp shouldBe exp
    }

    @ParameterizedTest
    @CsvSource(
        "1.0,   1",
        "42.0,  1",
    )
    fun `test set fractionalExp for full level`(fractionalExp: Double, level: Int) {
        levelSystem.exp = 1.0
        levelSystem.fractionalExp = fractionalExp

        levelSystem.level shouldBe level
        levelSystem.exp shouldBe 0
    }

    @ParameterizedTest
    @CsvSource(
        "0, 0.0,    0.0",
        "2, 12.0,   42.0",
    )
    fun `test get totalExp`(level: Int, exp: Double, totalExp: Double) {
        levelSystem.level = level
        levelSystem.exp = exp

        levelSystem.totalExp shouldBe totalExp
    }

    @ParameterizedTest
    @CsvSource(
        "-10.0, 0,  0.0",
        "0.0,   0,  0.0",
        "42.0,  2,  12.0",
    )
    fun `test set totalExp`(totalExp: Double, level: Int, exp: Double) {
        levelSystem.totalExp = totalExp

        levelSystem.level shouldBe level
        levelSystem.exp shouldBe exp
    }

    @ParameterizedTest
    @CsvSource(
        "15.0,  true",
        "5.0,   true",
        "18.0,  false",
    )
    fun `test hasExpTotal`(requiredExp: Double, has: Boolean) {
        levelSystem.level = 1
        levelSystem.exp = 5.0
        levelSystem.hasExpTotal(requiredExp) shouldBe has
    }

    @ParameterizedTest
    @CsvSource(
        "3.0,    10,     8.0",
        "105.0,  11,     0.0",
        "113.0,  11,     8.0",
    )
    fun `test giveExp`(expAmount: Double, level: Int, exp: Double) {
        levelSystem.level = 10
        levelSystem.exp = 5.0
        levelSystem.giveExp(expAmount)

        levelSystem.level shouldBe level
        levelSystem.exp shouldBe exp
    }

    @ParameterizedTest
    @CsvSource(
        "5.0,       10,     0.0",
        "6.0,       9,      99.0",
        "550.0,     0,      5.0",
        "555.0,     0,      0.0",
        "1000.0,    0,      0.0",
    )
    fun `test takeExp`(expAmount: Double, level: Int, exp: Double) {
        levelSystem.level = 10
        levelSystem.exp = 5.0
        levelSystem.takeExp(expAmount)

        levelSystem.level shouldBe level
        levelSystem.exp shouldBe exp
    }
}
